util/collections/src/garbage/atomic/counted.rs

```diff
@@ -61,15 +61,36 @@ impl GcCounter {
 		let slot = slot_timestamp % self.num_slots;
 		let (active_slot_timestamp, count) = &self.value_lifetimes[slot as usize];
 
-		// Atomically check and set the timestamp if it doesn't match current_time
+		// Atomically check and set the timestamp if it doesn't match the current slot
 		let active_slot = active_slot_timestamp.load(Ordering::Relaxed);
+		let active_amount = count.load(Ordering::Relaxed);
+
 		if active_slot == slot {
-			// Same timestamp, increment count
+			// Same timestamp, increment count.
+			// At worst, this adds to a later slot.
+			// But, such should not happen under safe usage of this API.
+			// Unless gc windows are very small.
 			count.fetch_add(amount, Ordering::SeqCst);
 		} else {
-			// Different timestamp, reset slot for new period
-			active_slot_timestamp.store(slot_timestamp, Ordering::SeqCst);
-			count.store(amount, Ordering::SeqCst);
+			// now we will zero out the slot and add the new value
+			// Use compare_exchange to safely reset the slot and count only
+			if active_slot_timestamp
+				.compare_exchange(active_slot, slot_timestamp, Ordering::SeqCst, Ordering::Relaxed)
+				.is_ok()
+			{
+				// Successfully updated the timestamp, now set the count only if the active_amount is unchanged (no one trying to push to the slot)
+				// This creates a condition wherein someone will win the race to set the first value and then everyone else will just add to it
+				if !count
+					.compare_exchange(active_amount, amount, Ordering::SeqCst, Ordering::Relaxed)
+					.is_ok()
+				{
+					// otherwise fetch add
+					count.fetch_add(amount, Ordering::SeqCst);
+				}
+			} else {
+				// If another thread updated the timestamp, the amount will also have been reset, so just go ahead and add
+				count.fetch_add(amount, Ordering::SeqCst);
```

Comment:
So the assumption here is that another thread has stored exactly the same `slot_timestamp` value into `active_slot_timestamp`. I guess this holds, unless a thread has been suspended for so long that the slot wrapped around. Which might happen e.g. if the machine was suspended for some reason, or someone's running a node on a laptop or similar.

Reply:
Yeah, so I read into this a bit more. Practically, atomic operations may involve invalidating cache lines on different cores. It's not quite such a nice use of an ISA's atomic instructions as I originally expected. 

So, we should expect synchronization to be on the order of nanoseconds. If a slot were indeed this small, we would get this wrap around. In that case, we're adding to a later slot. 

The problem would be the count could keep increasing without ever being collected under this condition. I'm going to scratch my head a little more to see if there's something clever. But, I think we may just have to say such a small slot size is not the intended usage.

Reply:
So, I think we should either leave this as is but increase the size of the buffer.

1. No matter how we play with these primitives, we're not going to have complete control over the critical section. If contended paths lead to `fetch_add` then we can increment past what was intended. If contended paths lead to `exchange`, then we will miss counts. In the current implementation, 
2. However, currently, we would only start to `gc` when we loop back around, which is also when we would start adding values again. We would want to actually make sure, with proper usage, we are leaving a gap to garbage collect on some factor of the `slot_duration`, the minimum being `1*slot_duration`. This means our buffer should, in fact, be strictly greater than `value_ttl.get() / gc_slot_duration.get();` and we kind of need to play with the mapping. We might instead want a notion of a relative "next" slot in the ring buffer instead of an absolute mapping. Or, we could use a double-length ring buffer.

Reply:
What I mean is, the OS does not guarantee a thread to be scheduled with any minimal time interval between run slots. So it's conceivable that a thread would run fetch_add on line 92 at a completely different time from another thread that has overwritten the same slot.

I think a way to be bulletproof here has to do with trying `compare_exchange` in a loop until the operation overwrites an expected value.